### PR TITLE
EFSOI implementation in fv3/v16

### PR DESCRIPTION
**Pull request by AndrewEichmann-NOAA**

### CODE REVIEW

File: scripts/exgdas_efsoi.sh [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+#      else
+#         if [ $DO_CALC_INCREMENT = "YES" ]; then
+#            #$NLN $COMOUT_ANL_ENS/$memchar/${APREFIX}atma00${FHR}${ASUFFIX}             sanl_${CDATE}_fhr0${FHR}_${memchar}
+#            $NLN $COMOUT_ANL_ENSFSOI/$memchar/${APREFIX}atma00${FHR}${ASUFFIX}             sanl_${CDATE}_fhr0${FHR}_ni${memchar}
+#         else
+#            #$NLN $COMOUT_ANL_ENS/$memchar/${APREFIX}atmi00${FHR}${ASUFFIX}             incr_${CDATE}_fhr0${FHR}_${memchar}
+#            $NLN $COMOUT_ANL_ENSFSOI/$memchar/${APREFIX}atmi00${FHR}${ASUFFIX}             incr_${CDATE}_fhr0${FHR}_ni${memchar}
+#         fi
+#      fi
+#   done
```
Comment (aerorahul) [inline scripts/exgdas_efsoi.sh:257]: why are commented lines kept here?

File: scripts/exgdas_efsoi.sh [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+# Cat runtime output files.
+cat stdout stderr > $COMOUT_ANL_ENSFSOI/$EFSOISTAT
+
+$NCP osense_${CDATE}.dat $COMOUT_ANL_ENSFSOI/$OSENSEOUT
+$NCP osense_${CDATE}.dat $OSENSE_SAVE_DIR/$OSENSEOUT
+
+################################################################################
+#  Postprocessing
+######## AFE remove after testing
+cp -r $DATA /scratch1/NCEPDEV/stmp4/Andrew.Eichmann/efsoi
```
Comment (aerorahul) [inline scripts/exgdas_efsoi.sh:467]: should not have hard-wired paths.

File: scripts/exgdas_efsoi.sh [hunk 3]
```diff
@@ -0,0 +1,475 @@
+#!/bin/ksh
```
Comment (aerorahul) [inline scripts/exgdas_efsoi.sh:1]: why can't we re-use `exgdas_enkf.sh`?  JGDAS_ENKF_EFSOI should be able to set the required variables to allow using that ex script.  Similar to the enkf_fcst script.

File: scripts/exgdas_efsoi_update.sh [hunk 4]
```diff
@@ -0,0 +1,413 @@
+#!/bin/ksh
```
Comment (aerorahul) [inline scripts/exgdas_efsoi_update.sh:1]: we now have 3 scripts;
exgdas_efsoi_update.sh, exgdas_efsoi.sh and the enkf_update script that are 90% same.

File: scripts/exgdas_efsoi_update.sh [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+# save for EFSOI task (still needed?)
+$NCP $COMOUT_ANL_ENS/$GBIASe $COMOUT_ANL_ENSFSOI
+
+# Cat runtime output files.
+cat stdout stderr > $COMOUT_ANL_ENSFSOI/$ENKFSTAT
+
+################################################################################
+#  Postprocessing
+######## AFE remove after testing
+#cp -r $DATA /scratch1/NCEPDEV/stmp4/Andrew.Eichmann/efsoi
```
Comment (aerorahul) [inline scripts/exgdas_efsoi_update.sh:405]: remove commented lines and blocks that are/were for debugging.

File: util/EFSOI_Utilities/scripts/copyefsoifiles.sh [hunk 6]
```diff
@@ -0,0 +1,107 @@
+#!/bin/sh -x
+
+
+EXPDIR=/scratch1/NCEPDEV/stmp4/Andrew.Eichmann/nov2019b
+STGDIR=/scratch1/NCEPDEV/stmp4/Andrew.Eichmann/efsoitest2/staging
```
Comment (aerorahul) [inline util/EFSOI_Utilities/scripts/copyefsoifiles.sh:5]: hard-wired paths.

File: util/EFSOI_Utilities/src/CMakeLists.txtORIG [hunk 7]
```diff
@@ -0,0 +1,29 @@
+cmake_minimum_required(VERSION 2.6)
```
Comment (aerorahul) [inline util/EFSOI_Utilities/src/CMakeLists.txtORIG:1]: This file should not be committed.

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 [hunk 8]
```diff
@@ -0,0 +1,927 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (aerorahul) [inline util/EFSOI_Utilities/src/gridio_efsoi.f90:439]: why this much space?

**Review by aerorahul — CHANGES_REQUESTED** [review]
This PR lacks adequate (or any) description.  What is added, what is modified?  What are the changes/additions in the PR?  What do the changes mean or do?

Above all, this is a monster PR.
There are changes or additions in JJOBS, ex scripts, enkf src code, enkf utilities, new utilities, python scripts, fix files.

I suggest breaking this PR into many smaller ones.
- enkf src code changes deserve its own PR.  Explain what they do.
- Are all these utilities added here used?  If so, add them in a separate PR explaining the use case and need for these utilities.
- There is duplication of ex level scripts.  I see 3 scripts that set up the enkf namelist and execute the enkf.  This adds burden of maintenance that should be avoided.  
- JJOBS need some information on what they do.  One line. 
- There are hard-wired paths in the scripts that need to be removed.

File: scripts/exgdas_efsoi_update.sh (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi_update.sh:1]: I can see how the two update scripts can/should be combined, but exgdas_efsoi.sh, while based on the update script and using the same namelist file, does run a different executable with different input and output files.

**Comment by AndrewEichmann-NOAA** [comment]
Rahul,

Thank you for the review. To address dividing up the monster PR, should this one be closed, and the other ones opened using their own branches from a forked master?

**Comment by aerorahul** [comment]
Yes.  You can close this PR.  From this branch, you could create separate branches with isolated and targeted changes.  Use those branches to issue their own PR's.

File: scripts/exgdas_efsoi.sh [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+################################################################################
+####  UNIX Script Documentation Block
+#                      .                                             .
+# Script name:         exgdas_efsoi.sh
+# Script description:  Make global_enkf update for efsoi
+#
+# Author:        Rahul Mahajan      Org: NCEP/EMC     Date: 2017-03-02
```
[the remark this answers is not in the document]
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:8]: fixed

File: scripts/exgdas_efsoi.sh [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+# Author:        Liaofan Lin/Andrew Eichmann          Date: 2020-12-03
+#
+# Abstract: This script runs the efsoi executable
+#
+# $Id$
+#
+# Attributes:
+#   Language: POSIX shell
+#   Machine: WCOSS-Cray/Theia
```
[the remark this answers is not in the document]
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:17]: fixed

File: scripts/exgdas_efsoi.sh [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+#      fname=$COMOUT_ANL_ENS/\${ftype}.ensmean
+#   else
+#      fname=$COMOUT_ANL_ENS/\$memchar/\$ftype
+#   fi
+#   tar -xvf \$fname
+#done
+#EOFuntar
+#   set -x
+#   chmod 755 $DATA/untar.sh
+#fi
```
[the remark this answers is not in the document]
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:199]: Removed. Apologies, I thought I'd taken out commented blocks

File: scripts/exgdas_efsoi_update.sh (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi_update.sh:1]: will reexamine

File: scripts/exgdas_efsoi_update.sh (hunk 5 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi_update.sh:405]: removed

File: util/EFSOI_Utilities/src/CMakeLists.txtORIG (hunk 7 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/CMakeLists.txtORIG:1]: removed

File: util/EFSOI_Utilities/src/gridio_efsoi.f90 (hunk 8 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/src/gridio_efsoi.f90:439]: removed

File: scripts/exgdas_efsoi.sh (hunk 1 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:257]: removed

File: scripts/exgdas_efsoi.sh (hunk 2 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:467]: removed

File: scripts/exgdas_efsoi.sh (hunk 3 above)
Comment (AndrewEichmann-NOAA) [reply scripts/exgdas_efsoi.sh:1]: will reexamine

File: jobs/JGDAS_EFSOI_FCST [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    cat $EFCSGRP
+    exit 99
+fi
+
+
+##############################################
+# Send Alerts
+##############################################
+if [ $SENDDBN = YES ] ; then
+   $DBNROOT/bin/dbn_alert MODEL ENKF1_MSC_fcsstat $job $EFCSGRP
```
[the remark this answers is not in the document]
Comment (AndrewEichmann-NOAA) [reply jobs/JGDAS_EFSOI_FCST:122]: removed

File: jobs/JGDAS_EFSOI [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+[[ $status -ne 0 ]] && exit $status
+
+
+
+##############################################
+# Send Alerts
+##############################################
+if [ $SENDDBN = YES  ] ; then
+   $DBNROOT/bin/dbn_alert MODEL ENKF1_MSC_enkfstat $job $COMOUT_ANL_ENSFSOI/${APREFIX}enkfstat
+fi
```
[the remark this answers is not in the document]
Comment (AndrewEichmann-NOAA) [reply jobs/JGDAS_EFSOI:115]: removed

File: util/EFSOI_Utilities/scripts/copyefsoifiles.sh (hunk 6 above)
Comment (AndrewEichmann-NOAA) [reply util/EFSOI_Utilities/scripts/copyefsoifiles.sh:5]: changed

**Comment by AndrewEichmann-NOAA** [comment]
closing, to be broken up into smaller PRs

**Pull request closed without merging**